### PR TITLE
fixes

### PR DESCRIPTION
**Pull request by Kimapr**
Fixes:
- all locked nodes not preventing non-owners from stealing stuff from them. (!!!)
- storage units and indexers nodes rendering pitch black (this might be a problem with my setup though)
- storage indexers shitting themselves on items with metadata

### CODE REVIEW

**Comment by loosewheel** [comment]
>all locked nodes not preventing non-owners from stealing stuff from them. (!!!)

The form should not display for a non-owner, but if there is some lag items could be moved before the form is replaced. Adding handlers to stop this introduces peculiar problems with hoppers. 

>storage units and indexers nodes rendering pitch black (this might be a problem with my setup though)

These nodes have the glasslike_framed draw type so they connect as a single unit. If this is only an issue on your setup you may need to change the draw type on yours, rather than change it for everyone.

>storage indexers shitting themselves on items with metadata

Basically did what you have in your PR. This worked on the items I tested, but they worked before. If you have the item that was having problems you may need to test it.

Some of the features (such as updating) on content db seem to have stopped working for me. You may need to manually re-install the mod.

**Comment by Kimapr** [comment]
> The form should not display for a non-owner, but if there is some lag items could be moved before the form is replaced. Adding handlers to stop this introduces peculiar problems with hoppers.

The form DOES actually display for a non-owner, but is replaced with another one shortly after. Even if it didn't display at all, cheat clients can interact with ANY inventory regardless of whether there is a formspec displaying it, so hiding the form is not a security measure. Another thing is that i have [a mod that lets you dig nodes with inventories even if not empty](https://gitlab.com/Kimapr/diggable_chests) and it does so by simulating the player taking all the items out of it and only then digging it. Since your mod allows anyone to take items from locked nodes, my mod is going to let anyone dig any locked node, along with everything in it.

> These nodes have the glasslike_framed draw type so they connect as a single unit. If this is only an issue on your setup you may need to change the draw type on yours, rather than change it for everyone.

This issue happened every time i tested it, and i couldn't find any other way to fix it. Might be an engine bug. What minetest version are you using? I'm on 5.6.0

> Basically did what you have in your PR. This worked on the items I tested, but they worked before. If you have the item that was having problems you may need to test it.

No, you didn't. My PR also escapes itemstrings before putting them into item_image_button, your fix doesn't. Because of that the formspec is going to break on items with metadata, like written books or technic chainsaws.

**Comment by loosewheel** [comment]
>No, you didn't. My PR also escapes itemstrings before putting them into item_image_button, your fix doesn't. Because of that the formspec is going to break on items with metadata, like written books or technic chainsaws.

You're right, I missed this. It shouldn't be v.item anyway. I've scraped that mechanism. Item strings in the formspec can make them very long. Now a memory list is maintained while the for is open and only the integer index into the form is used in the formspec.

Also, when a node is placed it only has a small blank form. When the owner right clicks to open it the form is updated (this already happens). When the owner dismisses the form the blank form is set again, unowned indexers don't. This way when a non-owner player opens an owned form they will only have access to the blank form before it updates. This has two caveats. If the owner currently has the form open when a non-owner player accesses it, because the full form is current, that is the form they will have access to until the form updates. Also, because the blank form is current when the owner opens it, that is the form they will see until it updates. I
